spec/support/custom_matchers.rb

```diff
@@ -3,6 +3,42 @@ module CustomMatchers
   def have_header(text)
     have_css 'h1', text: text
   end
+
+  # Returns a matcher for whether the given batch_item is in the batch's list of
+  # items. The select used is a data attribute containing the BatchItem ID.
+  def have_batch_item_row(batch_item)
+    raise_argument_error_if_not_batch_items batch_item
+    have_css("tr[data-batch-item-id=\"#{batch_item.id}\"]")
+  end
+
+  # Returns a matcher for whether all the given batch_items given are in the search
+  # results.
+  def have_batch_item_rows(batch_items = [])
+    raise_argument_error_if_not_batch_items batch_items
+    batch_items = batch_items.to_a.dup
+    batch_items.reduce have_batch_item_row(batch_items.shift) do |memo, batch_item|
+      memo.and(have_batch_item_row(batch_item))
+    end
+  end
+
+  # Returns a matcher for whether the given batch_items are the ONLY recrods in the
+  # serach results.
+  def only_have_batch_item_rows(batch_items = [])
+    raise_argument_error_if_not_batch_items batch_items
+    # Has all of the batch_items in the search results...
+    have_batch_item_rows(batch_items).and(
+      # ... and only has batch_items.count search results.
+      have_css("tr[data-batch-item-id]", count: batch_items.count)
+    )
+  end
+
+  def raise_argument_error_if_not_batch_items(batch_items = [])
```

Comment:
Ah.. i see what you're saying. I actually did that on purpose because I wanted the error to be thrown by the most immediate custom matcher to make it easier on devs. For instance, if you pass junk to `only_have_batch_item_rows` then the error message will indicate that `only_have_batch_item_rows` was used wrong.

Reply:
If we make it more efficient, and limit the call to `raise_argument_error_if_not_batch_items` to only the `have_batch_item_row` (which is the way I originally had it), then passing junk to `only_have_batch_item_rows` results in an error message referring to `have_batch_item_row`, which is not the matcher the developer is using.

It sacrifices calling `raise_argument_error_if_not_batch_items` more than needed in exchange for better clarity (IMO) to the developer.

I've been looking for a better way to do type checking on Ruby methods for a while, and haven't found anything that is awesome. I even checked out `Dry::Types` for this PR but I didn't see a very clean way to use it to coerce types on method arguments. But I want to read it again in.